go-controller/pkg/ovn/pods.go

```diff
@@ -497,3 +620,150 @@ func (oc *Controller) getPortAddresses(nodeName, portName string) (net.HardwareA
+			"of host network: %t and routing-network annotation: %s", pod.Name, pod.Spec.HostNetwork, routingNetworkAnnotation)
+		return nil
+	}
+
+	// if we found any gateways then we need to update current pods routing in the relevant namespace
+	if len(foundGws) == 0 {
+		klog.Warningf("No valid gateway IPs found for requested external gateway pod: %s", pod.Name)
+		return nil
+	}
+
+	for _, namespace := range strings.Split(routingNamespaceAnnotation, ",") {
+		nsInfo, err := oc.waitForNamespaceLocked(namespace)
+		if err != nil {
+			return err
+		}
+		defer nsInfo.Unlock()
+		nsInfo.routingExternalPodGWs[pod.Name] = foundGws
+		existingPods, err := oc.watchFactory.GetPods(namespace)
+		if err != nil {
+			return fmt.Errorf("failed to get all the pods for namespace %s, error: %v", namespace, err)
+		}
+		for _, gwIP := range foundGws {
+			for _, pod := range existingPods {
+				for _, podIP := range pod.Status.PodIPs {
+					var mask string
+					if utilnet.IsIPv6(net.ParseIP(podIP.IP)) {
+						mask = IPv6FullMask
+					} else {
+						mask = IPv4FullMask
+					}
+					gr := "GR_" + pod.Spec.NodeName
+					// TODO (trozet): use the go bindings here and batch commands
+					_, stderr, err := util.RunOVNNbctl("--", "--may-exist", "--policy=src-ip", "--ecmp",
+						"lr-route-add", gr, pod.Status.PodIP+mask, gwIP.String())
+					if err != nil {
+						klog.Errorf("Unable to add pod ecmp src-ip route to GR router, stderr:%q, err:%v", stderr, err)
+					} else {
+						klog.V(5).Infof("ECMP route added for pod: %s, on gr: %s, to gw: %s", pod.Name,
+							gr, gwIP.String())
+						if nsInfo.podExternalRoutes[podIP.IP] == nil {
+							nsInfo.podExternalRoutes[podIP.IP] = make(map[string]string)
+						}
+						nsInfo.podExternalRoutes[podIP.IP][gwIP.String()] = gr
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (oc *Controller) deletePodExternalGW(pod *kapi.Pod) {
+	routingNamespaceAnnotation := pod.Annotations[routingNamespaceAnnotation]
+	if routingNamespaceAnnotation == "" {
+		return
+	}
+	klog.Infof("External gateway pod: %s, detected for namespace(s) %s", pod.Name, routingNamespaceAnnotation)
+	for _, namespace := range strings.Split(routingNamespaceAnnotation, ",") {
+		nsInfo, err := oc.waitForNamespaceLocked(namespace)
+		if err != nil {
+			klog.Error(err)
+			continue
+		}
+		// check if any gateways were stored for this pod
+		foundGws := nsInfo.routingExternalPodGWs[pod.Name]
+		if foundGws == nil {
+			klog.Infof("No gateways found to remove for annotated gateway pod: %s on namespace: %s",
+				pod.Name, namespace)
+			nsInfo.Unlock()
+			continue
+		}
+
+		for _, gwIP := range foundGws {
+			// check for previously configured pod routes
+			for podIP, gwInfo := range nsInfo.podExternalRoutes {
+				if len(gwInfo) == 0 {
+					continue
+				}
+				gr := gwInfo[gwIP.String()]
+				if gr == "" {
+					continue
+				}
+				var mask string
+				if utilnet.IsIPv6(net.ParseIP(podIP)) {
```

Comment:
Yeah this really needs to be a separate function. That's like the 5th or 6th place it's used.

Reply:
done